### ISSUE TITLE
Predicting conversion rates

### ISSUE THREAD

**Issue by erikbern** [CamDavidsonPilon/lifelines#414]
Hi – this might be out of the scope of this library, but I've been working a bit on predicting conversion rates in the case where there is huge time lag. Turns out fitting a Weibull (or in some cases Gamma) distribution works quite well. However there's a crucial departure from survival analysis – modeling survival implies that "everyone dies in the long run" (the long-run conversion rate converges to 1.0) whereas when modeling conversion rates, it converges to some value 0 <= c <= 1. So there's one more parameter that I'm trying to fit for each cohort.

I've built a bunch of these models in Tensorflow but feels like I'm essentially reimplementing lifelines slowly. Wanted to see if it would make sense to do this inside lifelines or not. No strong feelings either way, happy to close this issue if this is completely out of the scope.

Code: https://github.com/better/convoys/blob/master/convoys/regression.py (this is highly janky – it's the first time I've used Tensorflow and there's a lot of code duplication)

**Comment by CamDavidsonPilon** [comment]
👋 @erikbern I've been following convoys actually. w.r.t. this issue, I do have some questions:

1. What does your modified Weibull model look like?

2. Lifelines doesn't have weibull-regression, so what you are adding is novel. It would be nice to eventually add to lifelines. I've recently been fighting with Python convergence of some models, and seeing you pivot to TF is definitely tempting me to try. However, I don't think I know enough yet to move to TF- I'd like to see performance metrics and ability to pull out the Hessian (I see you do grab it though).

**Comment by erikbern** [comment]
> 👋 @erikbern I've been following convoys actually. w.r.t. this issue, I do have some questions:

wow, shocked you found it!

> 
> What does your modified Weibull model look like?

the basic model is simple – the "CDF" is `c * (1 - exp(-(lambda*t)^k))`. It's not a real CDF because it obviously doesn't converge to 1.0

it gets a bit more complicated though... `c` is then modeled using a logistic regression (with unknown parameters `beta`) and `lambda` is the `exp(...)` of a linear combination of unknown parameters `alpha`). Fitting `k` as a single scalar to make it simple. Not sure if this model makes sense but seems to fit things decent for my datasets.

> 
> Lifelines doesn't have weibull-regression, so what you are adding is novel. It would be nice to eventually add to lifelines. I've recently been fighting with Python convergence of some models, and seeing you pivot to TF is definitely tempting me to try. However, I don't think I know enough yet to move to TF- I'd like to see performance metrics and ability to pull out the Hessian (I see you do grab it though).

Honestly the main reason I switched to TF was that I used to use `scipy.optimize` + `autograd` but for my gamma model i rely on `scipy.special.gammainc` which `autograd` doesn't support. I also like the idea of being able to use minibatches and GPU speedup. This is the first thing I've ever used TF for, but I've used Theano a fair amount in the past which I like.

Going back to the main question – is the distinction between survival analysis and conversion prediction something that you (or anyone) has given any thought? I've been surprised to see essentially no research on the latter, which is surprising given that it seems useful in practice

**Comment by CamDavidsonPilon** [comment]
I think I haven't given much thought to the distinction tbh - most of the models I use _don't_ go to zero (kaplan-meier for example), so I never really saw them as different problems. 

Putting my non-survival analysis hat on, I imagine there is a "beta regression" which would be appropriate for modelling conversion rates (quick Google search brings up [this paper](https://www.ime.usp.br/~sferrari/beta.pdf) - well cited and recent).

**Comment by erikbern** [comment]
> I think I haven't given much thought to the distinction tbh - most of the models I use don't go to zero (kaplan-meier for example), so I never really saw them as different problems.

I suspect using Kaplan-Meier for data that doesn't go to zero is not entirely correct and may introduce biases, but haven't spent too much time on it yet.

> 
> Putting my non-survival analysis hat on, I imagine there is a "beta regression" which would be appropriate for modelling conversion rates (quick Google search brings up this paper - well cited and recent).

The model in that paper doesn't deal with censored data and rates that vary over time, although it's possible you could basically take their model and multiply it with any model from survival analysis and fit both at the same time in some way.

Anyway, closing this issue for now.

**Comment by erikbern** [comment]
@CamDavidsonPilon I got convoys into a somewhat more useful state and wrote up some documentation: https://better.engineering/convoys/

Curious if you have any thoughts about this. I've found it super useful internally to analyze conversions. This is tricky for us since many of our users can take 6 or 9 months to transact. Weibull or Gamma distributions actually fit their conversions really beautifully and has let me extrapolate so that I can learn from recent cohorts much faster.
